lib/srv/app/watcher.go

```diff
@@ -82,7 +84,22 @@ func (s *Server) startResourceWatcher(ctx context.Context) (*services.AppWatcher
 		for {
 			select {
 			case apps := <-watcher.AppsC:
-				s.monitoredApps.setResources(apps)
+				appsWithAddr := make(types.Apps, 0, len(apps))
+				for _, app := range apps {
+					if app.GetPublicAddr() == "" {
+						c := app.Copy()
+						pubAddr, err := FindPublicAddr(s.c.AccessPoint, app.GetPublicAddr(), app.GetName())
+						if err == nil {
+							c.Spec.PublicAddr = pubAddr
+						} else {
+							s.log.WithError(err).Errorf("Unable to find public address for app %q, leaving empty.", app.GetName())
+						}
+						appsWithAddr = append(appsWithAddr, c)
+					} else {
+						appsWithAddr = append(appsWithAddr, app)
+					}
```

Comment:
Maybe something like this for less nested code?
```suggestion
					if app.GetPublicAddr() != "" {
						appsWithAddr = append(appsWithAddr, app)
						continue
					}
					c := app.Copy()
					pubAddr, err := FindPublicAddr(s.c.AccessPoint, app.GetPublicAddr(), app.GetName())
					c.Spec.PublicAddr = pubAddr
					if err != nil {
						s.log.WithError(err).Errorf("Unable to find public address for app %q, leaving empty.", app.GetName())
					}
					appsWithAddr = append(appsWithAddr, c)
```

Reply:
Thanks, I have extracted a method that (1) makes the loop cleaner (2) deals away with long if branch.

Reply:
Thanks, looks much better now!